gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/MaxLocalStep.java

```diff
@@ -34,23 +34,23 @@
  * @author Marko A. Rodriguez (http://markorodriguez.com)
  * @author Daniel Kuppitz (http://gremlin.guru)
  */
-public final class MaxLocalStep<E extends Comparable, S extends Iterable<E>> extends ScalarMapStep<S, E> {
+public final class MaxLocalStep<E extends Comparable, S extends Iterable<E>> extends OptionalScalarMapStep<S, E> {
 
     public MaxLocalStep(final Traversal.Admin traversal) {
         super(traversal);
     }
 
     @Override
-    protected E map(final Traverser.Admin<S> traverser) {
+    protected Optional<E> map(Traverser.Admin<S> traverser) {
+        Comparable result = null;
         final Iterator<E> iterator = IteratorUtils.asIterator(traverser.get());
         if (iterator.hasNext()) {
-            Comparable result = iterator.next();
+            result = iterator.next();
             while (iterator.hasNext()) {
                 result = max(iterator.next(), result);
             }
-            return (E) result;
         }
-        throw FastNoSuchElementException.instance();
+        return Optional.of((E) result);
```

Comment:
Seems this is throwing NPE. You need to use `ofNullable` instead.

Reply:
Unfortunately in TinkerPop, there is some case that returns `null` so I could not always treat like this and skip when the result is `null`. I reverted the change and extended just `MapStep`